Added CreditCardNumberValidator and Unit Testing for it. #10

Joaolfelicio opened the pull request:
#10 
Added CreditCardNumberValidator class.
Added CreditCardNumberValidator Unit Testing.
Updated readme to include the Credit Card Number Validator.

Joaolfelicio committed:
Added CreditCardNumberValidator and Unit Testing for it. #10

Joaolfelicio committed:
Updated class name from CreditCardValidator to CreditCardNumberValidator

afucher commented on src/Inquirer/Validators/CreditCardNumberValidator.cs:
```diff
@@ -0,0 +1,23 @@
+                .Replace(value, "");
+
+            // It matches a 16 digits number?
+            return new Regex(@"^\d{16}$")
```
According to this [wiki page](https://en.wikipedia.org/wiki/Payment_card_number) to be a valid credit card number we should use the Luhn algorithm. I found this [page](https://www.freeformatter.com/credit-card-number-generator-validator.html#validate) with some good explanation.

afucher reviewed (changes requested):
Thanks for your PR!

It seems that the credit card number validator is more complex than just check the numbers. Could you please review it?

Joaolfelicio commented:
Was not aware of that algorithm, will implement it. Thank you!

Joaolfelicio committed:
Implemented Luhn Algorithm, added more Unit Tests

Joaolfelicio committed:
Updated regex to only contain digits and to have a length of between 8 and 19 digits.

Credit Card Numbers are composed of 8 to 19 digits (Source: https://en.wikipedia.org/wiki/Payment_card_number).

Joaolfelicio committed:
Updated readme.

Updated error message for credit card validation.

Joaolfelicio committed:
Merge branch 'dev' into Feature_CreditCardValidator

afucher commented on test/InquirerUnitTest/Validators/CreditCardNumberValidatorUnitTest.cs:
```diff
@@ -0,0 +1,62 @@
+        [InlineData("2131")]
+        [InlineData("123456789012345'")]
+        [InlineData("asdfghjkl;qwerty")]
+        public void ShouldNotAcceptInvalidCreditCard(string validCreditCard)
```
As this validator has some complex rules, I think that break the test for invalid specific cases would be better. Like:
 - One for invalid size
 - One for invalid chars
 - One for invalid Luhn check

this would make the test more readable and easier to understand all the invalid cases

Joaolfelicio replied:
Good point, thank you for your feedback.
Working on it.

afucher reviewed (changes requested):
Thanks for implementing the algorithm!

Joaolfelicio committed:
Split unit testing methods by type of refusal.

Joaolfelicio committed:
Merge branch 'Feature_CreditCardValidator' of https://github.com/Joaolfelicio/Inquirer into Feature_CreditCardValidator

afucher commented on src/Inquirer/Validators/CreditCardNumberValidator.cs:
```diff
@@ -0,0 +1,81 @@
+            var isValidNumberFormat = new Regex(@"^[0-9]{8,19}$")
+                .Match(cleanCreditCardNumber).Success;
+
+            if (isValidNumberFormat == false)
+            {
+                return false;
+            }
+
+            return IsValidLuhnAlgorithm(cleanCreditCardNumber);
```
```suggestion
            if (isValidNumberFormat) return IsValidLuhnAlgorithm(cleanCreditCardNumber);
           
            return false;

            
```

afucher replied:
If we invert the code we can simplify it

Joaolfelicio replied:
Looks good to me, thanks!

afucher commented on test/InquirerUnitTest/Validators/CreditCardNumberValidatorUnitTest.cs:
```diff
@@ -0,0 +1,85 @@
+        [InlineData("111a-2222-3333-4444")]
+        [InlineData("1111/2222 3333 4444")]
+        [InlineData("123456789012345'")]
+        public void ShouldNotAcceptInvalidChars(string creditCardNumber)
```
Nice!! The tests are more meaningful now, thanks!

Joaolfelicio committed:
Inverted valid number format logic

Co-authored-by: Arthur Fücher <arthurfucher@gmail.com>

afucher reviewed (approved):
Thanks for your contribution! 🎉
